modules/migrate.py

```diff
@@ -0,0 +1,485 @@
+        soup = BeautifulSoup(lines)
+
+        # Non-modular Widget view page path relative to the content. This path
+        # will be the data attribute of any widget object tags.
+        widgetViewRelativePath = os.path.join(
+            os.path.relpath(widgetAbsolutePath, os.path.dirname(filePath)),
+            'index.html')
+
+        # Absolute path to the modular version of the widget view page.
+        absModuleViewPath = os.path.join(
+            modularWidgetAbsolutePath, 'index.html')
+        # Widget view page path relative to the content.
+        relativeModuleViewPath = os.path.relpath(
+            absModuleViewPath, os.path.dirname(filePath))
+
+        # Convert all object data (iframe src) links to new modular location.
+        # This migrates all the widgets to point to their modular versions, but
+        # leaves all parameters and JSON config files incorrectly relative to
+        # the wrong location.
+        lines = re.sub(r'(<object\s.*?data=")' + widgetViewRelativePath +'"',
+                       r'\g<1>' + relativeModuleViewPath + '"',
+                       lines,
+                       flags=re.IGNORECASE)
+
+        # For all object tags, clean up params and including config file. Use
+        # the old relative path because BeautifulSoup parse tree was created
+        # before the regex replace above and doesn't know about it.
+        for object in soup.find_all('object',
+                attrs={'data': widgetViewRelativePath}):
+            # Find all param tags for that widget
+            for param in object.find_all('param'):
+                # Param value could be anything, but it might be a relative
+                # path. Treat it as such creating an absolute path and see if it
+                # exists. If it does, assume it is a relative path to something
+                # in the project and update it to be relative to the new modular
+                # widget location.
+                val = param['value']
+                path = os.path.normpath(os.path.join(widgetAbsolutePath, val))
+
+                # NOTE(andy): For this test to work we have to check out the
+                # whole book. Sad times. But now we do.
+                if os.path.exists(path):
+                    newPath = os.path.relpath(path, modularWidgetAbsolutePath)
+
+                    # Target specifically the parameter with a relative path
+                    # inside the <object> tag we are migrating. If we didn't
+                    # include the <object> tag we could accidentally change
+                    # a relative path for a widget that wasn't migrating that
+                    # referenced the same asset / content.
+                    toReplace = (r'(<object.*?data="' + relativeModuleViewPath +
+                        '".*?<param.*?value=")' + val + '(".*?)(?=</object>)')
+                    lines = re.sub(toReplace,
+                                   r'\g<1>' + newPath + r'\g<2>',
```

Comment:
Same comment about extracting this regex to a variable named `replacement` for readability.